### PR TITLE
support upload of cookbooks using berkshelf

### PR DESCRIPTION
**Pull request by wjimenez5271**
support for Berkshelf to resolve and upload cookbooks

### CODE REVIEW

**Comment by wjimenez5271** [comment]
also see related: https://github.com/facebook/between-meals/pull/11

File: README.md [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # Grocery Delivery
 
+A fork of Facebook's [grocery delivery](https://github.com/facebook/grocery-delivery).
```
Comment (jaymzh) [inline README.md]: nope.

File: bin/grocery-delivery [hunk 2]
```diff
@@ -66,6 +66,12 @@ def read_checkpoint
 end
 
 def full_upload(knife)
+  if GroceryDelivery::Config.berks
+    puts 'berks!'
```
Comment (jaymzh) [inline bin/grocery-delivery]: debug leftover?

File: bin/grocery-delivery [hunk 3]
```diff
@@ -66,6 +66,12 @@ def read_checkpoint
...
+    GroceryDelivery::Log.warn('Uploading all cookbooks with berkshelf')
+    GroceryDelivery::Log.debug('using cookbook paths'+GroceryDelivery::Config.cookbook_paths.to_s)
```
Comment (jaymzh) [inline bin/grocery-delivery]: < 80 chars per line
Prefer string interpolation to concat in ruby, unless breaking lines.
Put spaces around operators.

File: bin/grocery-delivery [hunk 4]
```diff
@@ -66,6 +66,12 @@ def read_checkpoint
...
+    knife.berks_cookbook_upload_all()
```
Comment (jaymzh) [inline bin/grocery-delivery]: paryns not needed

File: bin/grocery-delivery [hunk 5]
```diff
@@ -66,6 +66,12 @@ def read_checkpoint
...
+  end
```
Comment (jaymzh) [inline bin/grocery-delivery]: don't you need an else around the other upload?

**Comment by jaymzh** [comment]
I've not used berkshelf - so I have a few questions that may be dumb.
- what about roles? I know jamie hates roles, but can you not have them at all in berkshelf?
- I feel like the berkshelf model and the grocery-delivery model were mutually-exclusive, can you talk me through the use-case a bit here? I'm not opposed to this if there's a good use-case, but I also don't want to set people up for failure.

**Comment by wjimenez5271** [comment]
Thanks for all the feedback. Those are great questions. I'll try my best to answer.

I have heard Jamie's thoughts on roles, and from what I can tell, his issue with them is their inherit ephemerality and lack of versioning. I share the same concern, but do like them for the ability to use them in knife searches. The alternative to roles is the "role-cookbook", however I don't think they have to be mutually exclusive. I plan to have a "chef role" for each role-cookbook, where all the logic and attribute data is in the role-cookbook, and the "chef role" is simply a pointer to that role-cookbook. This achieves the goals of the role-cookbook, while giving it some of the power of the native role object provided by Chef.

Re. your 2nd question, I thought so too, but the more I've explored this and sought after the opinions of others (there was a great hangops session on this that touched on it a bit), the less I've seen it as a strict dichotomy. The chef community embraces a loose model of how things are to be done--the thought being one-size-fits-all doesn't yield the best results. That being said, the use case for combining the two here is to manage community cookbooks in sane way. I think the monolithic repo has many benefits: it clearly describes a changes to an environment (branch = delta, instead of multiple branches for each repo involved), versioning can be based on git hash which enforces accurate "versioning" of cookbook code (instead of versions defined by the user which can mean different things if not managed carefully), just to name a few. But for community cookbooks, maintaining a duplicate copy in a local repo solely for the purposes of deployment ins't ideal. Using Berkshelf to resolve these dependencies addresses this problem well, while not compromising the benefits we get from our own assets being in a single repo. We're not making any changes to the comm. cookbooks, if we need to set any attributes or override anything, we'll use wrapper cookbooks. Therefore our changes are confined to the domain of our repo, again supporting the goal that change-sets can be described with a single branch/pull request/commit.

Hope that helps explain the use case, and thanks again for considering this idea!

**Comment by jaymzh** [comment]
Thanks for the explanation. I *think* I get it. So the workflow is something like:
* Community Cookbooks are manually put in berkshelf
* Roles and Your Cookbooks are managed by GD where...
  * Roles are pushed directly and continuously to Chef
  * Cookbooks are fed to berkshelf
* Berkshelf then feeds cookbooks to Chef on some sort of basis

Something like:

```
roles ---------------------------> grocery-delivery ----------> Chef Server
community cookbooks ----------------------------> berkshelf-----^
your cookbooks -------------> grocery-delivery ------^
```
?

If you do this - especially given that you have roles that are pointers to cookbooks, how do you keep roles in sync with cookbooks since berkshelf will block things going to Chef but roles are instantly in Chef?

**Comment by wjimenez5271** [comment]
Berkshelf should push cookbooks to Chef instantly as well I think, it just makes sure dependencies are pushed also. Perhaps there is a situation I'm not thinking of?

**Comment by jaymzh** [comment]
From Jamie's talk, that wasn't the impression I got. You had to "promote" them using some sort of test or manual process? As I understand it the whole point is that you can "safely" upload to berkshelf regularly without it immediately going to your chef server, so you can take a whole bundle of cookbooks, test them and then put that entire regression-tested set of cookbooks into prod at once.

**Comment by wjimenez5271** [comment]
Hmm, can't say I've seen that, possibly that was the model in v 2.0 of Berkshelf as I know they made a lot of changes in the latest version (3.0). I'll see if I can get some more info/confirmation of the expected behavior of berks.

**Comment by sciurus** [comment]
@jaymzh there is a distinction between `berks install`, which copies a cookbook into your $HOME/.berkshelf/cookbooks, and `berks upload`, which uploads it to your chef server. Maybe that is what you were thinking of? In Berkshelf I'm not aware of anything that behaves quite like you describe. I think the "berkshelf way" to put a set of cookbooks into prod at once would be to upload them, then edit the version constraints in the chef environment for prod.

I use Berkshelf the same way that @wjimenez5271 does. It's purely for managing dependencies. I don't use environments or cookbook versioning at all. I have a homegrown program that runs in CI to upload changes from my single chef repo to the chef server. That program also runs Berkshelf to upload the community cookbooks I depend on to the chef server. Now that I'm aware of grocery-delivery, I may look into using it instead. Having it handle Berkshelf for me would be a plus.

**Comment by jaymzh** [comment]
Excellent. Thanks for the explanation. Sounds awesome. OK, I'm onboard with the change.

We just need to clean up the stuff from the code review, and you need to sign the CLA... I don't know why the Facebook bot didn't bug you... I'll have a poke at that.

**Comment by jaymzh** [comment]
Here's a link to the CLA: https://code.facebook.com/cla - @wjimenez5271 - you'll need to sign it before we can merge your changes.

**Comment by wjimenez5271** [comment]
Thanks! CLA completed.

File: bin/grocery-delivery [hunk 6]
```diff
@@ -66,8 +66,15 @@ def read_checkpoint
 end
 
 def full_upload(knife)
-  GroceryDelivery::Log.warn('Uploading all cookbooks')
-  knife.cookbook_upload_all
+  if GroceryDelivery::Config.berks
+    GroceryDelivery::Log.warn('Uploading all cookbooks with berkshelf')
+    GroceryDelivery::Log.debug("using cookbook paths:
```
Comment (jaymzh) [inline bin/grocery-delivery]: don't leave the newline in the string, do:

```
    GroceryDelivery::Log.debug(
      "using cookbook paths:  #{GroceryDelivery::Config.cookbook_paths.to_s}"
    )
```

(If you had to further break it, then our style says to use string concatenation to cross lines)

File: bin/grocery-delivery [hunk 7]
```diff
@@ -66,8 +66,15 @@ def read_checkpoint
 end
 
 def full_upload(knife)
-  GroceryDelivery::Log.warn('Uploading all cookbooks')
-  knife.cookbook_upload_all
+  if GroceryDelivery::Config.berks
+    GroceryDelivery::Log.warn('Uploading all cookbooks with berkshelf')
+    GroceryDelivery::Log.debug("using cookbook paths: 
+                                #{GroceryDelivery::Config.cookbook_paths.to_s}")
```
Comment (jaymzh) [inline bin/grocery-delivery]: sticking it in `#{}` will automatically call .to_s on it

File: bin/grocery-delivery [hunk 8]
```diff
@@ -120,7 +127,12 @@ def partial_upload(knife, repo, checkpoint, local_head)
   end
 
   knife.cookbook_delete(deleted_cookbooks) if deleted_cookbooks
```
Comment (jaymzh) [inline bin/grocery-delivery]: is this still the right place to do deletes - or do you need to tell berks you want to delete to handle any dependencies?

File: grocery_delivery.gemspec [hunk 9]
```diff
@@ -1,6 +1,6 @@
 Gem::Specification.new do |s|
   s.name = 'grocery_delivery'
-  s.version = '0.0.5'
+  s.version = '0.0.6'
```
Comment (jaymzh) [inline grocery_delivery.gemspec]: don't change that, we'll change it when we release.

File: bin/grocery-delivery (hunk 8 above)
Comment (wjimenez5271) [inline bin/grocery-delivery]: `berks` doesn't have a notion of deleting cookbooks, so don't think we can do anything with that. Perhaps they'll add that feature down the line.

File: bin/grocery-delivery [hunk 10]
```diff
@@ -66,8 +66,15 @@ def read_checkpoint
 end
 
 def full_upload(knife)
-  GroceryDelivery::Log.warn('Uploading all cookbooks')
-  knife.cookbook_upload_all
+  if GroceryDelivery::Config.berks
+    GroceryDelivery::Log.warn('Uploading all cookbooks with berkshelf')
+    GroceryDelivery::Log.debug(
```
Comment (jaymzh) [inline bin/grocery-delivery]: style nit:
* If the opening paren gets it's own line so does the closing
* indentation is two spaces:

```
    GroceryDelivery::Log.debug(
      "using cookbook paths: #{GroceryDelivery::Config.cookbook_paths}"
    )
    knife.berks_cookbook_upload_all
  else
```

**Comment by jaymzh** [comment]
Your dependent change to BM has been merged. Minor style here to fix up and then I'll go ahead and merge, looks good. Thanks for contributing and for all your work during the review process!

File: bin/grocery-delivery (hunk 10 above)
Comment (wjimenez5271) [inline bin/grocery-delivery]: makes sense, amended last commit

**Comment by wjimenez5271** [comment]
Awesome, thanks for considering this PR!

**Pull request merged**